vscode/extension.js

```diff
@@ -0,0 +1,349 @@
+        PLUGIN_ID = "udp://" + addr.address + ":" + addr.port;
+        console.log("UDP Server listening on", PLUGIN_ID);
+    };
+    
+    message (msg, rinfo) {
+        let data = JSON.parse(msg.toString());
+        this.handleSuggestion(data);
+    };
+    
+    error (error) {
+        console.log("UDP Server error:", error);
+    };
+    
+    close () {
+        console.log("Closing");
+    };
+    
+    // Kite will send back suggestions. Depending on its type,
+    // we handle it in one of three different ways: 
+    // an 'apply':
+    //      - applies the suggested fix onto the active editor.
+    // a 'highlight':
+    //      - instructs the editor to highlight a range of code.
+    // a 'clear':
+    //      - instructs the editor to clear any highlights added
+    //      to the editor.
+    handleSuggestion (suggestion) {
+        let suggestionType = suggestion['type'];
+        if (suggestionType === 'apply') {
+            this.handleApply(suggestion);
+        } else if (suggestionType === 'highlight') {
+            this.handleHighlight(suggestion);
+        } else if (suggestionType === 'clear') {
+            this.handleClear(suggestion);
+        }
+    };
+    
+    // Applies a suggested fix onto the currently active editor.
+    handleApply (suggestion) {
+        if (!this.validBuffer(suggestion)) {
+            return;
+        }
+        
+        const editor = vscode.window.activeTextEditor;
+        
+        if (!editor) {
+            return;
+        }
+        
+        for (var i = 0; suggestion.diffs && i < suggestion.diffs.length; i++) {
+            let diff = suggestion.diffs[i];            
+            editor.edit(function (editBuilder) {
+                let startPosition = new vscode.Position(diff.linenum - 1, 0);
+                let endPosition = new vscode.Position(diff.linenum - 1, diff.line_src.length);
+                let removedRange = new vscode.Range(startPosition, endPosition);
+                editBuilder.replace(removedRange, diff.line_dest.trimRight());
+            });
+        }
+        
+        this.handleClear();
+    };
+        
+    // Adds a highlight.
+    handleHighlight (suggestion) {
+        if (!this.validBuffer(suggestion)) {
+            return;
+        }
+        
+        const editor = vscode.window.activeTextEditor;
+        
+        if (!editor) {
+            return;
+        }
+        
+        let allHighlights = [];
+        let documentUri = editor.document.uri;
+        let documentText = editor.document.getText();
+        
+        for (var i = 0; suggestion.diffs && i < suggestion.diffs.length; i++) {
+           let diff = suggestion.diffs[i];
+           
+           // Build the start and end position of the highlight before adding it.
+           let startPosition = this.buildHighlightPosition(diff.begin, diff.linenum, documentText);
+           let endPosition = this.buildHighlightPosition(diff.end, diff.linenum, documentText);
```

Comment:
@dhung09 If diffs are guaranteed to only apply to a single line, then yes - I think this could work. I guess my reasoning was around the edge case in which a diff could wrap.